prereq_tools/base.py

```diff
@@ -64,6 +64,151 @@
 else:
     import configparser as ConfigParser
 
+class EnvironmentModule():
+    """Class for utilizing Modules component to load environment modules"""
+
+    def __init__(self):
+        """Load Modules for initializing envirables"""
+        self._module_func = self._init_module_func(lambda *x: False)
+        self._module_load = self._init_mpi_module()
+
+    def set_module_func(self, func):
+        """Set the mpi module func"""
+        self._module_func = func
+
+    @staticmethod
+    def _module_old(*args):
+        """invoke module and save environment"""
+        cmd = ['/usr/bin/modulecmd', 'python']
+
+        for arg in args:
+            cmd.append(arg)
+        output = subprocess.check_output(cmd)
+        exec(output)
+
+    def _init_module_func(self, default_func):
+        """Initialize environment modules"""
+        init_file = None
+        for root, _dirs, files in os.walk('/usr/share/Modules'):
+            for fname in files:
+                if fname == "python.py":
+                    temp = os.path.join(root, fname)
+                    if "init" in temp:
+                        init_file = temp
+                        break
+        if init_file is None:
+            return default_func
+
+        try:
+            subprocess.check_call(['sh', '-l', '-c', 'module -V'],
+                                  stdout=DEVNULL, stderr=DEVNULL)
+        except subprocess.CalledProcessError:
+            return self._module_old
+
+        try:
+            exec(open(init_file).read())
+        except KeyError:
+            return default_func
+
+        # pylint: disable=undefined-variable
+        return module
+        # pylint: enable=undefined-variable
+
+    def _init_mpi_module(self):
+        """init mpi module function"""
+        try:
+            subprocess.check_call(['sh', '-l', '-c', 'module -V'],
+                                  stdout=DEVNULL, stderr=DEVNULL)
+        except subprocess.CalledProcessError:
+            # older version of module return -1
+            return self._mpi_module_old
+
+        return self._mpi_module
+
+    def _mpi_module(self, mpi):
+        """attempt to load the requested module"""
+        mpich = ['mpi/mpich-x86_64', 'mpich']
+        openmpi = ['mpi/openmpi3-x86_64', 'gnu-openmpi',
+                   'mpi/openmpi-x86_64']
+        if mpi == "mpich":
+            load = mpich
+            unload = openmpi
+        else:
+            load = openmpi
+            unload = mpich
+
+        for to_load in load:
+            if self._module_func('is-loaded', to_load):
+                return True
+
+        for to_unload in unload:
+            if self._module_func('is-loaded', to_unload):
+                self._module_func('unload', to_unload)
+
+        for to_load in load:
+            if self._module_func('load', to_load):
+                print("Loaded %s" % to_load)
+                return True
```

Comment:
why not just first unload everything in unload and then load only what you need?

Reply:
could do that, I guess.  I'm sick of this patch right now

Reply:
The main reason I did this is to preserve user environment